execute/tokendata/observer_background.go

```diff
@@ -0,0 +1,431 @@
+}
+
+// IsTokenSupported simply forwards the call to the underlying observer.
+func (o *backgroundObserver) IsTokenSupported(
+	sourceChain cciptypes.ChainSelector, msgToken cciptypes.RampTokenAmount) bool {
+	return o.observer.IsTokenSupported(sourceChain, msgToken)
+}
+
+// Close stops the background goroutines and cleans up resources.
+func (o *backgroundObserver) Close() {
+	close(o.done)
+	o.wg.Wait()
+}
+
+// startWorkers starts the worker goroutines that process messages from the queue.
+func (o *backgroundObserver) startWorkers() {
+	o.lggr.Info("waiting for existing (if any) workers to stop")
+	o.wg.Wait()
+	o.lggr.Info("all workers stopped, new workers are starting...")
+
+	for i := 0; i < o.numWorkers; i++ {
+		o.wg.Add(1)
+		workerID := i
+		go o.worker(workerID)
+	}
+}
+
+// worker is a goroutine that processes messages from the queue.
+// It will stop using the o.done channel.
+func (o *backgroundObserver) worker(id int) {
+	lggr := logger.With(o.lggr, "workerID", id)
+	lggr.Info("worker started")
+
+	defer o.wg.Done()
+	for {
+		select {
+		case <-o.done:
+			lggr.Info("worker stopped after receiving done signal")
+			return
+		case <-o.msgQueue.newMsgSignalChan:
+			lggr.Debug("new job signal received")
+
+			msg, ok := o.msgQueue.dequeue()
+			if !ok {
+				lggr.Debug("nothing to work on, waiting for new job signal")
+				continue
+			}
+
+			lggr := logger.With(lggr,
+				"msgID", msg.Header.MessageID.String(),
+				"sourceChain", msg.Header.SourceChainSelector.String(),
+				"seqNum", msg.Header.SequenceNumber.String(),
+				"numTokens", len(msg.TokenAmounts),
+			)
+			lggr.Infow("processing message")
+
+			// observe only this single message and use a timeout for the observation
+			observationTimeoutCtx, cancel := context.WithTimeout(context.Background(), o.observeTimeout)
+			tokenData, err := o.observer.Observe(observationTimeoutCtx,
+				map[cciptypes.ChainSelector]map[cciptypes.SeqNum]cciptypes.Message{
+					msg.Header.SourceChainSelector: {msg.Header.SequenceNumber: msg},
+				})
+			cancel()
+
+			if err != nil {
+				lggr.Errorw("message observation failed, message pushed again to the queue", "err", err)
+				o.msgQueue.enqueue(msg, o.reprocessInterval)
+				continue
+			}
+
+			if _, chainExists := tokenData[msg.Header.SourceChainSelector]; !chainExists {
+				lggr.Errorw("underlying observer did not return token data for the chain")
+				o.msgQueue.enqueue(msg, o.reprocessInterval)
+				continue
+			}
+
+			if _, seqExists := tokenData[msg.Header.SourceChainSelector][msg.Header.SequenceNumber]; !seqExists {
+				lggr.Errorw("underlying observer did not return token data for the sequence number")
+				o.msgQueue.enqueue(msg, o.reprocessInterval)
+				continue
+			}
+
+			if !tokenData[msg.Header.SourceChainSelector][msg.Header.SequenceNumber].SupportedAreReady() {
+				lggr.Infow("token data not ready by the underlying observer, message pushed again to the queue")
+				o.msgQueue.enqueue(msg, o.reprocessInterval)
```

Comment:
How many times are we going to do this? I'm worried that retrying indefinitely some faulty messages might impact processing of "valid" ones

Reply:
Could we let the plugin dictate retries by waiting until the next time it asks for the token data?